lib/Data/Chronicle/Writer.pm

```diff
@@ -128,38 +127,119 @@ publish "category1::name1" in Redis if C<publish_on_set> is true.
-    my $archive  = shift // 1;
-    my $ttl      = shift // $self->ttl;
+    my $self         = shift;
+    my $category     = shift;
+    my $name         = shift;
+    my $value        = shift;
+    my $rec_date     = shift;
+    my $archive      = shift // 1;
+    my $suppress_pub = shift // 0;
+    my $ttl          = shift // $self->ttl;
 
-    die "Recorded date is undefined" unless $rec_date;
-    die "Recorded date is not a Date::Utility object" if ref $rec_date ne 'Date::Utility';
-    die "Cannot store undefined values in Chronicle!" unless defined $value;
-    die "You can only store hash-ref or array-ref in Chronicle!" unless (ref $value eq 'ARRAY' or ref $value eq 'HASH');
+    $self->mset([[$category, $name, $value]], $rec_date, $archive, $suppress_pub, $ttl);
+
+    return 1;
+}
+
+=head2 mset
 
-    $value = JSON::MaybeXS->new->encode($value);
+Example:
+
+    $chronicle_writer->mset([["category1", "name1", $value1], ["category2, "name2", $value2], ...]);
+
+Store a piece of data "value1" under key "category1::name1", "category2::name2", etc in Pg and Redis. Will
+publish "category1::name1", "category2::name2", etc in Redis if C<publish_on_set> is true.
+
+=cut
+
+sub mset {
+    my $self         = shift;
+    my $entries      = shift;
+    my $rec_date     = shift;
+    my $archive      = shift // 1;
+    my $suppress_pub = shift // 0;
+    my $ttl          = shift // $self->ttl;
+
+    $self->_validate_value($_->[2]) foreach @$entries;
+    $self->_validate_rec_date($rec_date);
 
-    my $key    = $category . '::' . $name;
     my $writer = $self->cache_writer;
 
     # publish & set in transaction
     $writer->multi;
-    my $encoded = encode_utf8($value);
-    $writer->publish($key, $encoded) if $self->publish_on_set;
-    $writer->set(
-        $key => $encoded,
-        $ttl ? ('EX' => $ttl) : ());
-    $writer->exec;
+    foreach my $entry (@$entries) {
+        my $category = $entry->[0];
+        my $name     = $entry->[1];
+        my $value    = $entry->[2];
+
+        my $key = $self->_generate_key($category, $name);
+
+        my $encoded = encode_json_utf8($value);
+        $writer->publish($key, $encoded) if $self->publish_on_set && !$suppress_pub;
+        $writer->set(
+            $key => $encoded,
+            $ttl ? ('EX' => $ttl) : ());
 
-    $self->_archive($category, $name, $value, $rec_date) if $archive and $self->dbic;
+        $self->_archive($category, $name, $encoded, $rec_date) if $archive and $self->dbic;
+    }
+    $writer->exec;
 
     return 1;
 }
 
+=head2 subscribe
+
+Example:
+
+    $chronicle_writer->subscribe("category1", "name1", $code_ref);
+
+=cut
+
+sub subscribe {
+    my ($self, $category, $name, $subref) = @_;
+    die 'publish_on_set must be enabled to subscribe' unless $self->publish_on_set;
+    die 'Subscription requires a coderef' if ref $subref ne 'CODE';
+
+    my $key = $self->_generate_key($category, $name);
+    return $self->cache_writer->subscribe($key, $subref);
```

Comment:
I admit I'm having difficulty working out what type a `cache_writer` would be here - it doesn't seem to be documented in this file what that would be. Is there a better explanation somewhere?

But if it's going to be something Redis-related (perhaps using `Net::Async::Redis`), there's a problem in that a subscribing connection cannot also be used for regular data updates. If subscriptions are required, you'll have to maintain two connections.

Reply:
In our case it's an instance of `RedisDB`.

This is actually an issue that should probably get properly resolved in this package, because I think even the original developer didn't have it completely straight.

https://github.com/binary-com/perl-Data-Chronicle/blob/77853aa1c0bde788885edad2c8507875b9ed8d10/lib/Data/Chronicle/Reader.pm#L131
In the `Reader` package the codes checks if `cache_reader` is a blessed object or not, and treats it accordingly. Where `Writer` happily calls `multi`, `publish`, `set`, and `exec` on `cache_writer` with no such check.

I will remove this ambiguity from the code.